### DISCUSSION TITLE
How to complie Candle for Native devices (Android, IOS)?

### DISCUSSION THREAD

**Discussion by akashicMarga** [huggingface/candle#2081] [General]
How can we build candle binaries which can be used in Android or IOS projects?

**Comment by soupslurpr** [comment]
I used UniFFI to generate Kotlin bindings to my Rust project and then call the necessary methods from Kotlin. 
Additionally, in .cargo/config.toml I put:

[target.aarch64-linux-android]
ar = "C:/Users/user/AppData/Local/Android/Sdk/ndk/26.1.10909125/toolchains/llvm/prebuilt/windows-x86_64/bin/llvm-ar.exe"
linker = "C:/Users/user/AppData/Local/Android/Sdk/ndk/26.1.10909125/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android23-clang.cmd"

[target.x86_64-linux-android]
ar = "C:/Users/user/AppData/Local/Android/Sdk/ndk/26.1.10909125/toolchains/llvm/prebuilt/windows-x86_64/bin/llvm-ar.exe"
linker = "C:/Users/user/AppData/Local/Android/Sdk/ndk/26.1.10909125/toolchains/llvm/prebuilt/windows-x86_64/bin/x86_64-linux-android23-clang.cmd"

[env]
CC_aarch64-linux-android = "C:/Users/user/AppData/Local/Android/Sdk/ndk/26.1.10909125/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android23-clang.cmd"
CC_x86_64-linux-android = "C:/Users/user/AppData/Local/Android/Sdk/ndk/26.1.10909125/toolchains/llvm/prebuilt/windows-x86_64/bin/x86_64-linux-android23-clang.cmd"
AR_aarch64_linux_android = "C:/Users/user/AppData/Local/Android/Sdk/ndk/26.1.10909125/toolchains/llvm/prebuilt/windows-x86_64/bin/llvm-ar.exe"
AR_x86_64_linux_android = "C:/Users/user/AppData/Local/Android/Sdk/ndk/26.1.10909125/toolchains/llvm/prebuilt/windows-x86_64/bin/llvm-ar.exe"


Adjust for where the ndk is for you and the installed version.


Finally, I also added tokenizers with default-features = false and features = ["onig"] so OpenSSL isn't compiled.

**Reply by akashicMarga** [reply]
I have made all changes in config but still getting open-ssl error.

**Reply by soupslurpr** [reply]
Try doing `cargo tree` to see what's pulling in openssl

**Reply by akashicMarga** [reply]
cargo tree is not showing any lib pulling openssl. My cargo.toml has:

`[dependencies]
accelerate-src = { version = "0.3.2", optional = true }
anyhow = "1.0.79"
candle = { git = "https://github.com/huggingface/candle", package = "candle-core" }
candle-nn = { git = "https://github.com/huggingface/candle" }
candle-transformers = { git = "https://github.com/huggingface/candle" }
clap = { version = "4.2.4", features = ["derive"] }
hf-hub = { version = "0.3.2", features = ["tokio"] }
intel-mkl-src = { version = "0.8.1", optional = true }
safetensors = "0.4.2"
serde = { version = "1.0.196", features = ["derive"] }
serde_json = "1.0.113"
tokenizers = { version = "0.19.0", default-features = false, features = ["onig"] }
tracing = "0.1.40"
tracing-chrome = "0.7.1"
tracing-subscriber = "0.3.18"
byteorder = "1.4.3"
rand = "0.8.5"
cpal = "0.15.2"`

**Reply by soupslurpr** [reply]
isn't it hf-hub?

**Reply by akashicMarga** [reply]
no i don't see openssl in hf-hub with cargo tree.

**Reply by akashicMarga** [reply]
@soupslurpr  it seems like hf-hub  had dependency of openssl although it did not show in cargo tree but I made a blank cargo project with same toml and removing hf-hub made openssl error go. But now unable to compile `gemm-f16`

**Reply by soupslurpr** [reply]
What version of the NDK are you using? You might have to use 26 or higher.

**Reply by akashicMarga** [reply]
i am using 26.1.10909125 , same one that you mentioned. Maybe there might be something with the host platform that I have which is M series MacBook. 

/Users/akashsingh/Library/Android/sdk/ndk/26.1.10909125/toolchains/llvm/prebuilt/darwin-x86_64/bin/

**Reply by soupslurpr** [reply]
What error did you get with gemm-f16? I think I got the same and fixed it by adding this to config.toml:

rustflags = [
    "-C", "target-feature=+fp16",
]

**Reply by akashicMarga** [reply]
ok the test project works now after adding this flag. How did you use hf-hub? or you have downloaded the model and then using it directly without compiling hf-hubapi.

**Reply by soupslurpr** [reply]
Oh great! Just predownloaded and loaded the model without compiling hf-hub.

**Reply by akashicMarga** [reply]
![c](https://github.com/huggingface/candle/assets/18519731/233a4c1a-dbf7-42b3-9ec7-fe4ae068b3ee)
![d](https://github.com/huggingface/candle/assets/18519731/b0848f3c-83bd-4924-b18f-bd1f61adbe3a)


One pic showing performance of bert and another performance of phi-2. Phi-2 performance is slower than llama.cpp which give 2.6 token/sec and candle is giving 1.6 token/sec. Is there any optimisation which needs to be done wrt to android target?

**Reply by soupslurpr** [reply]
Are you using the quantized model with llama.cpp maybe?

**Reply by akashicMarga** [reply]
> Are you using the quantized model with llama.cpp maybe?

Yes

**Reply by soupslurpr** [reply]
And you're running the quantized one with Candle too?

**Reply by akashicMarga** [reply]
> And you're running the quantized one with Candle too?

Yes same 4 bit quantised.

**Reply by soupslurpr** [reply]
Ah I think it's just that Candle doesn't have all the optimizations https://github.com/huggingface/candle/blob/cfab6e761696c18b1ce5d3a339ab57ef191ca749/candle-core/src/quantized/neon.rs
